eclass/rocm.eclass

```diff
@@ -0,0 +1,278 @@
+# Copyright 2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# @ECLASS: rocm.eclass
+# @MAINTAINER:
+# Gentoo Science Project <sci@gentoo.org>
+# @AUTHOR:
+# Yiyang Wu <xgreenlandforwyy@gmail.com>
+# @SUPPORTED_EAPIS: 7 8
+# @BLURB: Common functions and variables for ROCm packages written in HIP
+# @DESCRIPTION:
+# ROCm packages such as sci-libs/<roc|hip>* can utilize functions in this
+# eclass.  Currently, it handles the AMDGPU_TARGETS variable via USE_EXPAND, so
+# user can use USE flag to control which GPU architecture to compile, and
+# ensure coherence among dependencies. It also specify CXX=hipcc, to let hipcc
+# compile. Another important function is src_test, which checks whether a valid
+# KFD device exists for testing, and then execute the test program.
+#
+# Most ROCm packages use cmake as build system, so this eclass does not export
+# phase functions which overwrites the phase functions in cmake.eclass. Ebuild
+# should explicitly call rocm_src_* in src_configure and src_test.
+#
+# @EXAMPLE:
+# # Example for ROCm packages in https://github.com/ROCmSoftwarePlatform
+# @CODE
+# inherit cmake rocm
+# SRC_URI="https://github.com/ROCmSoftwarePlatform/${PN}/archive/rocm-${PV}.tar.gz -> ${P}.tar.gz"
+# SLOT="0/$(ver_cut 1-2)"
+# IUSE="test"
+# REQUIRED_USE="${ROCM_REQUIRED_USE}"
+# RESTRICT="!test? ( test )"
+#
+# RDEPEND="
+#     dev-util/hip
+#     sci-libs/rocBLAS:${SLOT}[${ROCM_USEDEP}]
+# "
+#
+# S=${WORKDIR}/${PN}-rocm-${PV}
+#
+# src_configure() {
+#     local mycmakeargs=(
+#         -DBUILD_CLIENTS_TESTS=$(usex test ON OFF)
+#     )
+#     rocm_src_configure
+# }
+#
+# src_test() {
+#     rocm_src_test
+# }
+# @CODE
+#
+# # Example for packages depend on ROCm libraries -- a package depend on
+# # rocBLAS, and use comma seperated ${HCC_AMDGPU_TARGET} to determine GPU
+# # architecture to compile. Requires ROCm version >5.
```

Comment:
The example above also depended on rocblas… or was that a mistake?

Reply:
> Overall, I don't really understand the purpose of this eclass. Reading it, it feels like you're trying hard to combine two different eclasses into one.

The purpose of this eclass is to provide common funcitons in ROCm and ROCm related packages. So there are two types of ebuilds that can use rocm.eclass, and this is why there are two examples:

1. ROCm math libraries in https://github.com/ROCmSoftwarePlatform/, developed by AMD. currently they all use cmake, have similar `SRC_URI`, share some cmake arguments, and are common in testing, so `rocm_src_configure` and `rocm_src_test` is designed for them. `USE_EXPAND` and `get_amdgpu_flags` is used to control the GPU architecture being compiled.
2. Other libraries utilizing ROCm, like CuPy, PyTorch, Tensorflow, etc. They have their own build system, so `rocm_src_configure` and `rocm_src_test` is useless. But controlling which architecture to compile is useful, so they can still use rocm.eclass, to make dependency coherent on GPU architectures.

So, rocm.eclass should be used with cmake.eclass in occasion 1., and used with whatever-build-system in occasion 2. Exporting `src_test` and `src_configure` would overwrite other cmake phase funcitons, so I did not export them.

Reply:
> The example above also depended on rocblas… or was that a mistake?

I should pointed out the first example also depend on rocBLAS. I'll add that